### ISSUE TITLE
NeoPic: PNG transparent color treated as solid in some cases

### ISSUE THREAD

**Issue by PiKeyAr** [defisym/OpenFusionExamples#17]
Hi again! I found this small issue loading animations from PNG files exported from Fusion and thought it might be a bug. Here are the PNG files I'm trying to load:
[jump.zip](https://github.com/defisym/OpenFusionExamples/files/11078946/jump.zip)

**Type of issue**
Extension bug

**Describe the bug**
Some PNG files with a background transparent color load with a solid background. This can happen with files exported from Fusion's Picture Editor.

**To Reproduce**
1. In Fusion's Picture Editor, create an image with a transparent background (do not create the alpha channel).
2. Export the image as PNG.
3. Load the image in NeoPic.
4. Instead of being transparent, the background is colored as the color that was supposed to be transparent.

**Expected behavior**
Transparent background when the file is loaded.

**Screenshots**
![image](https://user-images.githubusercontent.com/23312549/227947923-5c790293-135a-449a-9231-86b72ecfe5cd.png)

Left: PNG file loaded ingame; right: same PNG file opened in Paint.net

**Desktop (please complete the following information):**
 - OS: Windows 11 22H2

**Comment by PiKeyAr** [comment]
Thank you for the response. I think the issue is that Fusion doesn't account for cases when a specific color is set as transparent, which is possible in PNG24 without the alpha channel. So they make the user pick the transparent color manually with Active Picture etc. It's also odd that Photoshop doesn't read it properly, it displays fine in other viewers/editors and Windows Explorer.

Thank you also for the workaround. For single pictures this sounds like a solution. In my case I'm afraid I can't use it because I'm loading animation frames from a lib object, and the animations are updated almost every frame. Anyway, it's no big deal because I can just resave the images with a proper alpha channel.

Appreciate all your work on this repo!

**Comment by defisym** [comment]
> Thank you also for the workaround. For single pictures this sounds like a solution. In my case I'm afraid I can't use it because I'm loading animation frames from a lib object, and the animations are updated almost every frame. Anyway, it's no big deal because I can just resave the images with a proper alpha channel.

This object uses fusion's load API, so it cannot process this type of file automatically. 
Just note that all changes done by display object could be save back to the lib by using the load from pointer / load from display feature. You can add this to your load event. 
If the copy of surface is bottom neck then I can add a move feature, which moves display object's resource to lib without copying, and attach that object to lib. 

Thanks for your reply.

**Comment by defisym** [comment]
Hi, the latest build added a load callback to make this process easier

![image](https://user-images.githubusercontent.com/55148235/233795412-a975868c-8903-4f1e-af3b-35c82142f5c9.png)

BTW, you can also check the animation system:

![image](https://user-images.githubusercontent.com/55148235/233795499-8df7cb29-c258-426e-a021-f5d5dc792970.png)

https://user-images.githubusercontent.com/55148235/233795478-d048bdea-0242-4959-990c-a3cf942e5b6a.mp4

**Comment by PiKeyAr** [comment]
Hi! Sorry for the late response. I was able to test the "Set transparent color" actions, and it worked fine. Though, I think I should stick to loading PNG32 files with a real alpha channel anyway because there is probably no performance benefit to PNG24, and because I need to update graphics almost every frame for animated characters.

Thank you also for the animation system. This looks excellent, I am going to use it for effects.

Sorry for the offtopic, but if you don't mind another question: is it possible for the Lib object to cache all PNG files in a folder recursively? I tried the "Set Preload Path" action but the lib size reported in the debugger remains at 0 and increases when the "Load From Lib" action is used. Also, sometimes the display object reports "From Lib: False" even though the image was loaded previously using the "Load From Lib" action. Admittedly I have a lot of small files to load, about 180 PNGs for different character animations.

**Comment by defisym** [comment]
>  I tried the "Set Preload Path" action but the lib size reported in the debugger remains at 0 and increases when the "Load From Lib" action is used.

Preload series will load in background and merge to main lib once complete, and `on preload complete` is triggered then.
Once you switch to another frame or destroy preload lib, then unmerged files will be released, as `rdPtr` passed to it is dangling now.
Note that due to fusion's limitation child process can only load to bitmaps so this may crash your app if you reach the 3GB memory limit

> Also, sometimes the display object reports "From Lib: False" even though the image was loaded previously using the "Load From Lib" action

Once you do some actions like blur & change transparent color, the object will first clone current surface then detach from lib to keep ones in lib untouched. The latest version has actions works like `std::move` to let you attach to lib again instead of traditional way (lib: load from object, object: load from lib), which has one unnecessary clone.

> Admittedly I have a lot of small files to load, about 180 PNGs for different character animations.

for animation system, you only need to give relative path in Json file, and extension will auto load the actual file to lib you set as animation source, by the base path & key you given together.

**Comment by PiKeyAr** [comment]
Sorry for the super late response, and thanks for the explanation! I should've mentioned that I'm using a custom animation system that I made before the animation features were added to this object, so I preload animation frames with the library and use another object to display them manually.


The issue with "From lib: False" was indeed caused by manipulation (I was using the image offset functionality), I removed that and it is working fine now. I think this can be closed now.
